### PR TITLE
prov/verbs/RDM: fix a crash of the CM thread that occurs if the allocation of the PD fails

### PR DESCRIPTION
**Pull request by gladkovdmitry17**
This patch fixes the crash of the verbs/RDM's CM thread when the allocation of the protection domain fails.
This occurs due to thread is keeping alive when `rdm_cm` is freed by `free(_domain->rdm_cm);` in the end of the error handling of the `fi_domain` function.
The problem is being fixed by extended error handling in the verbs' `fi_domain` fucntion

Signed-off-by: Dmitry Gladkov <dmitry.gladkov@intel.com>

### CODE REVIEW

**Comment by gladkovdmitry17** [comment]
The fabtest's test results are the following:
```
$ date; ./runfabtests.sh -p ./ -c 192.168.65.1 -s 192.168.65.1 verbs | tee test.out
Mon Oct  2 19:58:19 MSK 2017
# Test                                                  Result
# --------------------------------------------------------------
fi_getinfo_test -p "verbs":                             Pass
fi_av_test -g 192.168.10.1 -n 1 -s 192.168.65.1 -p "verbs":      Pass
fi_dom_test -n 2 -p "verbs":                            Pass
fi_eq_test -p "verbs":                                  Pass
fi_cq_test -p "verbs":                                  Pass
fi_mr_test -p "verbs":                                  Pass
fi_cntr_test -p "verbs":                                Pass
fi_dgram g00n13s -p "verbs":                            Pass
fi_rdm g00n13s -p "verbs":                              Pass
fi_msg g00n13s -p "verbs":                              Pass
fi_cm_data -p "verbs":                                  Pass
fi_cq_data -p "verbs":                                  Pass
fi_dgram -p "verbs":                                  Notrun
fi_dgram_waitset -p "verbs":                          Notrun
fi_msg -p "verbs":                                      Pass
fi_msg_epoll -p "verbs":                                Pass
fi_msg_sockets -p "verbs":                              Pass
fi_poll -t queue -p "verbs":                          Notrun
fi_poll -t counter -p "verbs":                        Notrun
fi_rdm -p "verbs":                                      Pass
fi_rdm_rma_simple -p "verbs":                         Notrun
fi_rdm_rma_trigger -p "verbs":                        Notrun
fi_shared_ctx -p "verbs":                             Notrun
fi_shared_ctx --no-tx-shared-ctx -p "verbs":          Notrun
fi_shared_ctx --no-rx-shared-ctx -p "verbs":          Notrun
fi_shared_ctx -e msg -p "verbs":                      Notrun
fi_shared_ctx -e msg --no-tx-shared-ctx -p "verbs":      Pass
fi_shared_ctx -e msg --no-rx-shared-ctx -p "verbs":    Notrun
fi_shared_ctx -e dgram -p "verbs":                    Notrun
fi_shared_ctx -e dgram --no-tx-shared-ctx -p "verbs":    Notrun
fi_shared_ctx -e dgram --no-rx-shared-ctx -p "verbs":    Notrun
fi_rdm_tagged_peek -p "verbs":                          Pass
fi_scalable_ep -p "verbs":                            Notrun
fi_cmatose -p "verbs":                                  Pass
fi_rdm_shared_av -p "verbs":                          Notrun
fi_multi_mr -e msg -V -p "verbs":                     Notrun
fi_multi_mr -e rdm -V -p "verbs":                     Notrun
fi_recv_cancel -e rdm -V -p "verbs":                  Notrun
fi_unexpected_msg -e msg -i 10 -p "verbs":            Notrun
fi_unexpected_msg -e rdm -i 10 -p "verbs":            Notrun
fi_unexpected_msg -e dgram -i 10 -p "verbs":          Notrun
fi_unexpected_msg -e msg -S -i 10 -p "verbs":         Notrun
fi_unexpected_msg -e rdm -S -i 10 -p "verbs":         Notrun
fi_unexpected_msg -e dgram -S -i 10 -p "verbs":       Notrun
fi_msg_pingpong -I 5 -p "verbs":                        Pass
fi_msg_pingpong -I 5 -v -p "verbs":                     Pass
fi_msg_bw -I 5 -p "verbs":                              Pass
fi_msg_bw -I 5 -v -p "verbs":                           Pass
fi_rma_bw -e msg -o write -I 5 -p "verbs":            Notrun
fi_rma_bw -e msg -o read -I 5 -p "verbs":             Notrun
fi_rma_bw -e msg -o writedata -I 5 -p "verbs":          Pass
fi_rma_bw -e rdm -o write -I 5 -p "verbs":              Pass
fi_rma_bw -e rdm -o read -I 5 -p "verbs":               Pass
fi_rma_bw -e rdm -o writedata -I 5 -p "verbs":          Fail
fi_msg_rma -o write -I 5 -p "verbs":                  Notrun
fi_msg_rma -o read -I 5 -p "verbs":                   Notrun
fi_msg_rma -o writedata -I 5 -p "verbs":                Pass
fi_msg_stream -I 5 -p "verbs":                          Pass
fi_rdm_atomic -I 5 -o all -p "verbs":                 Notrun
fi_rdm_cntr_pingpong -I 5 -p "verbs":                 Notrun
fi_rdm_multi_recv -I 5 -p "verbs":                      Pass
fi_rdm_pingpong -I 5 -p "verbs":                        Pass
fi_rdm_pingpong -I 5 -v -p "verbs":                     Pass
fi_rdm_rma -o write -I 5 -p "verbs":                    Pass
fi_rdm_rma -o read -I 5 -p "verbs":                     Pass
fi_rdm_rma -o writedata -I 5 -p "verbs":                Pass
fi_rdm_tagged_pingpong -I 5 -p "verbs":                 Pass
fi_rdm_tagged_pingpong -I 5 -v -p "verbs":              Pass
fi_rdm_tagged_bw -I 5 -p "verbs":                       Pass
fi_rdm_tagged_bw -I 5 -v -p "verbs":                    Pass
fi_dgram_pingpong -I 5 -p "verbs":                    Notrun
fi_rc_pingpong -n 5 -p "verbs":                         Pass
fi_rc_pingpong -n 5 -e -p "verbs":                      Pass
# --------------------------------------------------------------
# Total Pass                                                40
# Total Notrun                                              32
# Total Fail                                                 1
# Percentage of Pass                                        97
# --------------------------------------------------------------
```

The failure of the `fi_rma_bw -e rdm -o writedata -I 5 -p "verbs"` is known issue, becasue the verbs provider doesn't support writedata operation.

**Comment by gladkovdmitry17** [comment]
@a-ilango, Could you take a look, please?

File: prov/verbs/src/verbs_domain.c [hunk 1]
```diff
@@ -423,9 +424,17 @@ fi_ibv_domain(struct fid_fabric *fabric, struct fi_info *info,
...
-err3:
+err5:
 	if (_domain->rdm)
 		rdma_destroy_event_channel(_domain->rdm_cm->ec);
+err4:
+	if (ibv_dealloc_pd(_domain->pd))
+		VERBS_INFO_ERRNO(FI_LOG_DOMAIN, "ibv_dealloc_pd", errno);
+err3:
+	if (_domain->rdm) {
+		_domain->rdm_cm->fi_ibv_rdm_tagged_cm_progress_running = 1;
```
Comment (a-ilango) [inline prov/verbs/src/verbs_domain.c]: Shouldn't this be set to 0?

Comment (gladkovdmitry17) [reply prov/verbs/src/verbs_domain.c]: Sure, it should be set to 0.

The original changes that fix the reported problem in the PR #3268. Wrong copy-pasting :(
Thank you for being attentive.

**Pull request merged**